modules/product/src/Form/AddToCartForm.php

```diff
@@ -150,9 +164,20 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
     if (!$cart) {
       $cart = $this->cartProvider->createCart('default', $store);
     }
-    $quantity = $form_state->getValue('quantity');
+
     $combine = $form['#settings']['combine'];
-    $this->cartManager->addEntity($cart, $variation, $quantity, $combine);
+
+    $line_item = $this->cartManager->createLineItem($variation);
+    $line_item_form_display = EntityFormDisplay::collectRenderDisplay($line_item, self::LINE_ITEM_FORM_DISPLAY);
+
+    $line_item_values = $line_item_form_display->extractFormValues($line_item, $form, $form_state);
```

Comment:
AFAIK this will populate the fields on $line_item, you don't need to call ->set() yourself afterwards.